### PR TITLE
Create Unit-test-instructions.md and ensure CI runs unit tests

### LINKED ISSUE #155: Add frontend unit testing capabilities to repo
1. Make sure people are able to write and run unit tests in Jest for both FE and BE 
2. Make sure the continuous integration automatically runs the unit tests.

- [x] Introduction
- [x] Frontend setup

### LINKED ISSUE #217: Review W9 documentations
- [x] team minutes
- [x] Client minutes
- [x] Client agenda
- [x] Unit test instructions

### PR DESCRIPTION
**Pull request by wrc9817**

### CODE REVIEW

**Review by aleksssk — APPROVED** [review]
Looks good

**Comment by wrc9817** [comment]
if anything goes wrong, just discard your changes. It is somehow not working on my computer but should not be a big deal.

File: documentation/tech-directives/unit-test-instructions.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# Introduction #
+- The purpose of creating and running Unit test is to check functionality and quality of the code from each part of the software.
+
+# Requirement #
+  - Each **core** functional class are detected by a Unit Test Class.
+  - Within Unit test class, Each core method from the detected class should be attached with **at least two** test Unit Tests     with different given input and expected result.
+  - All the Unit Tests should be **commented** to indicate the detected methods.
```
Comment (jpabadir) [inline documentation/tech-directives/unit-test-instructions.md:7]: Is this a standard? I would tend to be against over-commenting. It is often clear which function is being tested, in a unit test.

File: documentation/tech-directives/unit-test-instructions.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  - To guarantee the functionality, all the Unit tests should be **passed**.
+  - The coverage should be **at least 80 percent** or higher.
+
+# Setup Jest #
+  ### Frontend ###
+  1. Install Jest: run the following command in a terminal: `npm install --save-dev jest@26.6.0`
+  2. Install React test renderer: run the following command in a terminal when you are done #1: `npm install reac-test-            renderer`
```
Comment (jpabadir) [inline documentation/tech-directives/unit-test-instructions.md]: Minor: extra space in `npm install reac-test-            renderer`

Also, "react" is missing a t

File: documentation/tech-directives/unit-test-instructions.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+       expect(sum(1, 2)).toBe(3);
+     });
+     ```
+     When you are done writing tests, run `npm run test` and the following result should print:
+     
+     ```javascript
+     PASS  ./sum.test.js
+     ✓ adds 1 + 2 to equal 3 (5ms)
+     ```
+  ### Backend ###
```
Comment (jpabadir) [inline documentation/tech-directives/unit-test-instructions.md]: Can you also do the backend instructions?

**Review by jpabadir — CHANGES_REQUESTED** [review]
This is good work, with a couple minor changes / points of discussion + missing back-end instructions.

File: documentation/tech-directives/unit-test-instructions.md (hunk 1 above)
Comment (wrc9817) [reply documentation/tech-directives/unit-test-instructions.md:7]: Yea That's what I meant. The standard for testing twice for each is what my prof asked me to do in my previous assignments. We can just do once for each method.

File: documentation/tech-directives/unit-test-instructions.md (hunk 2 above)
Comment (wrc9817) [reply documentation/tech-directives/unit-test-instructions.md]: Corrected. TY!

**Review by katmartel — APPROVED** [review]
Looks good, and it looks like you incorporated the suggested modifications, so great work!

File: documentation/tech-directives/unit-test-instructions.md (hunk 1 above)
Comment (jpabadir) [reply documentation/tech-directives/unit-test-instructions.md:7]: Sorry, I agree with the two tests for each method. What I disagree with is: 
"All the Unit Tests should be **commented** to indicate the detected methods.". I think it is often extraneous to comment tests, unless they're complex. What do you think @wrc9817 ?

File: .github/workflows/link.yml [hunk 4]
```diff
@@ -17,6 +17,12 @@ jobs:
 
     # Steps represent a sequence of tasks that will be executed as part of the job
     steps:
+      - uses: actions/checkout@v1
+      - name: Run Tests
+        run: |
+          npm install
+          npm test
+
```
Comment (jpabadir) [inline .github/workflows/link.yml]: Does this work? As per our Discord conversation, I thought this didn't work. If not, remove it from the PR to get my approval.

**Comment by jpabadir** [comment]
Is issue #217 one you made for yourself to keep track of reviews you have to do? If so, why is it linked to this PR?

**Comment by wrc9817** [comment]
> Is issue #217 one you made for yourself to keep track of reviews you have to do? If so, why is it linked to this PR? 

Because I also need to review all the documentations I made and this instruction is included.

**Comment by wrc9817** [comment]
> > Is issue #217 one you made for yourself to keep track of reviews you have to do? If so, why is it linked to this PR? 
> 
> Because I also need to review all the documentations I made and this instruction is included.

And if this PR is merged, which means i'm done reviewing.

File: documentation/tech-directives/unit-test-instructions.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  - To guarantee the functionality, all the Unit tests should be **passed**.
+  - The coverage should be **at least 80 percent** or higher.
+
+# Setup Jest #
+  1. Install Jest: run the following command in a terminal: 
+      `npm install --save-dev jest@26.6.0`
+  2. Install React test renderer: run the following command in a terminal when you are done #1: 
+      `npm install react-test-renderer`
+  3. Start server: 
+      run `cd src/front-end/ ` or  `cd src/back-end/ ` and then `npm start`
```
Comment (jpabadir) [inline documentation/tech-directives/unit-test-instructions.md]: npm start does not run on the backend

File: documentation/tech-directives/unit-test-instructions.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    
+     In *sum.test.js*:
+     
+     ```javascript
+     const sum = require('./sum');
+     test('adds 1 + 2 to equal 3', () => {
+       expect(sum(1, 2)).toBe(3);
+     });
+     ```
+     When you are done writing tests, run `npm run test` and the similar result should print:
```
Comment (jpabadir) [inline documentation/tech-directives/unit-test-instructions.md:38]: This does not work on the back end

**Review by jpabadir — CHANGES_REQUESTED** [review]
2 comments about the backend instructions + https://github.com/jpabadir/pace-c/pull/203/files#r517268330

File: .github/workflows/link.yml (hunk 4 above)
Comment (jpabadir) [reply .github/workflows/link.yml]: Update : as per our Discord conversation, I can have a look at whether or not this works.

File: documentation/tech-directives/unit-test-instructions.md (hunk 1 above)
Comment (wrc9817) [reply documentation/tech-directives/unit-test-instructions.md:7]: I meant that just comment on which function or method you are testing.

Comment (jpabadir) [reply documentation/tech-directives/unit-test-instructions.md:7]: I understand. I think it's extraneous to describe which method we're testing if we call the tests something descriptive, such as testBuildMyArray().

Comment (jpabadir) [reply documentation/tech-directives/unit-test-instructions.md:7]: I understand if you disagree, though.

File: .github/workflows/link.yml [hunk 7]
```diff
@@ -16,7 +16,7 @@ jobs:
     runs-on: ubuntu-latest
 
     # Steps represent a sequence of tasks that will be executed as part of the job
-    steps:
+    steps:
```
Comment (jpabadir) [inline .github/workflows/link.yml]: Can you get rid of this?

**Review by jpabadir — APPROVED** [review]
You can merge, but please, before you do so: 

1. Address my comment on the link.yml file
2. Rename the issue you're fixing to something to describe what you actually did in this PR (get rid of the CI part in the issue title, and make it clear this only tackles front-end
3. Make a new issue to say we need to do the CI part and the backend. You can self-assign if you want to own this.

File: documentation/tech-directives/unit-test-instructions.md (hunk 1 above)
Comment (wrc9817) [reply documentation/tech-directives/unit-test-instructions.md:7]: I'm okay with both. Or we can probably ask what aleks and katreenar think.

**Pull request merged**